kong/init.lua

```diff
@@ -347,57 +535,80 @@ function Kong.balancer()
     end
   end
 
-  current_try.ip   = addr.ip
-  current_try.port = addr.port
+  local ip, port = addr.ip, addr.port
+
+  current_try.ip   = ip
+  current_try.port = port
+
+  if mock then
+    ip, port = mock_addr()
+  end
 
   -- set the targets as resolved
   ngx_log(ngx_DEBUG, "setting address (try ", addr.try_count, "): ",
-                     addr.ip, ":", addr.port)
-  local ok, err = set_current_peer(addr.ip, addr.port)
+    ip, ":", port)
+
+  local ok, err = set_current_peer(ip, port)
   if not ok then
     ngx_log(ngx_ERR, "failed to set the current peer (address: ",
-            tostring(addr.ip), " port: ", tostring(addr.port),"): ",
-            tostring(err))
+      tostring(ip), " port: ", tostring(port),"): ",
+      tostring(err))
     return ngx.exit(500)
   end
 
   ok, err = set_timeouts(addr.connect_timeout / 1000,
-                         addr.send_timeout / 1000,
-                         addr.read_timeout / 1000)
+   addr.send_timeout / 1000,
+    addr.read_timeout / 1000)
   if not ok then
     ngx_log(ngx_ERR, "could not set upstream timeouts: ", err)
   end
 
-  core.balancer.after()
+  core.balancer.after(ctx, mock)
 end
 
-function Kong.rewrite()
+function Kong.rewrite(mock)
   local ctx = ngx.ctx
-  core.rewrite.before(ctx)
+  core.rewrite.before(ctx, mock)
 
-  -- we're just using the iterator, as in this rewrite phase no consumer nor
-  -- api will have been identified, hence we'll just be executing the global
-  -- plugins
-  for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins, true) do
-    plugin.handler:rewrite(plugin_conf)
+  if mock then
+    mock_plugins(ctx)
+  else
+    -- we're just using the iterator, as in this rewrite phase no consumer nor
+    -- api will have been identified, hence we'll just be executing the global
+    -- plugins
+    for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins, true) do
+      plugin.handler:rewrite(plugin_conf)
+    end
   end
 
-  core.rewrite.after(ctx)
+  core.rewrite.after(ctx, mock)
 end
 
-function Kong.access()
+function Kong.access(mock)
   local ctx = ngx.ctx
 
-  core.access.before(ctx)
+  core.access.before(ctx, mock)
 
   ctx.delay_response = true
 
-  for plugin, plugin_conf in plugins_iterator(singletons.loaded_plugins, true) do
-    if not ctx.delayed_response then
-      local err = coroutine.wrap(plugin.handler.access)(plugin.handler, plugin_conf)
-      if err then
-        ctx.delay_response = false
-        return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
+  if mock then
```

Comment:
Having all these extra ifs on all phases worries me a little bit. If we could do the check once at the beginning and then use separate functions I think I would prefer it (this would probably go against @Tieske's recommendation regarding code duplication). I strongly recommend some sort of performance analysis if we end up merging this change as it is now.

Reply:
Yes, that's what I argued, but having single boolean checks is not maybe that bad. Of course this doesn't prevent things totally not going out of sync. E.g. someone may forgot to add mock check and then it could run some code that was not supposed to run (the first attempt secured us against that, but it duplicated some code).